equinox/nn/_attention.py

```diff
@@ -308,6 +380,39 @@ def __call__(
                     "process_heads must not change the shape of the heads."
                 )
 
+        if state is None:
+            causal_mask_offset = 0
+        else:
+            if self.kv_cache is None:
+                raise ValueError(
+                    "State was provided, but cannot use autoregressive decoding "
+                    "without specifying "
+                    "`MultiheadAttention(..., kv_cache=...)`. "
+                    "See `equinox.nn.StandardKVCache` for an example."
+                )
+
+            key_state, value_state, index, state = self.kv_cache(
+                state, key_heads, value_heads
+            )
+            causal_mask_offset = index - kv_seq_length
+            key_heads = key_state
+            value_heads = value_state
+            kv_seq_length = self.state_length
```

Comment:
I like this! It helps to make the `MultiheadAttention.__call__` code a little less of a gigantic monolith.